docker/configure_workers_and_start.py

```diff
@@ -389,29 +554,219 @@ def generate_base_homeserver_config() -> None:
     subprocess.run(["/usr/local/bin/python", "/start.py", "migrate_config"], check=True)
 
 
+def parse_worker_types_from_env(
+    requested_worker_types: List[str],
+) -> Dict[str, Dict[str, Any]]:
+    """Read the desired list of worker from environment variables and prepare the data
+    for use in generating worker config files while also checking for potential gotchas.
+
+    Args:
+        requested_worker_types: The string pulled from the environment containing
+            the worker requested data.
+
+    Returns: A  dict of all information needed to generate worker files. Format:
+        {'worker_name':
+            {'worker_base_name': 'base_name'},
+
+            {'worker_types_set':
+                Set{'worker_type', 'other_worker_type'}
+            }
+        }
+    """
+    # Checking performed:
+    #   1. If a requested name contains a space
+    #   2. If a requested name contains either kind of quote mark
+    #   3. If a requested name ends with a digit
+
+    # A counter of worker_type -> int. Used for determining the name for a given
+    # worker type(s) when generating its config file, as each worker's name is just
+    # worker_(name|type(s)) + instance #
+    worker_type_counter: Dict[str, int] = defaultdict(int)
+
+    # Similar to above, but more finely grained. This is used to determine we don't have
+    # more than a single worker for cases where multiples would be bad(e.g. presence).
+    worker_type_shard_counter: Dict[str, int] = defaultdict(int)
+
+    # Dict of worker name's. This is used to check that a name requested doesn't clash
+    # with an existing name in the context of a differing worker_type, as it will error
+    # with 'Address in use'(e.g. "to_device, to_device=typing" would not work).
+    # Follows the pattern:
+    # ["worker_name": "worker_type(s)"]
+    worker_name_checklist: Dict[str, str] = {}
+
+    # The final result of all this processing
+    dict_to_return: Dict[str, Any] = {}
+
+    # Handle any multipliers requested for a given worker.
+    multiple_processed_worker_types = apply_requested_multiplier_for_worker(
+        requested_worker_types
+    )
+
+    # Check each requested worker for a requested name
+    for worker_type_string in multiple_processed_worker_types:
+        # Shortcut these to avoid processing and skip an 'else' block when no worker
+        # name is actually requested.
```

Comment:
Good idea, I'll update that. I was just calling out the fact that I was shortcutting the processing of the `if` statement to avoid writing out an `else`. I think I use that styling at least one other time. I call it 'running the gauntlet', declare the values then they just have to survive.

Reply:
it definitely has its place, but possibly loses it somewhat if you have to explain it in a comment :-). Sometimes the `else` can be clearer, ... probably a bit subjective though

Reply:
Removed comment in b8741d743